1600761 - Visual/Color improvements to Treeherder apps

yogmel opened the pull request:
## Description of issue
Some color combinations does not pass the [WCAG contrast](https://developer.mozilla.org/pt-BR/docs/Web/Accessibility/Understanding_WCAG/Perceivable/Color_contrast?utm_source=devtools&utm_medium=a11y-panel-checks-color-contrast) recommendation.

Also, for color blindness distinction, texts with red and green should be accompanied by an icon.

## Strategy
Using _Firefox Accessibility Inspector_'s beta "Check for issues", I've checked for "**_Contrast_**" issues. That was done for each view and a different commit was made. Then, it was a manual color tweaking of each issue. 

Some colors became harder to distinguish because of the contrast rate.  That was seem mostly in blue and green colors. Maybe propose a **different color hue**?

## Testing
Those were mostly visual and using Accessibility and Inspector devtools tabs.

yogmel commented:
As for the icons support, there are a couple that need them, IMO. All of them are in Treeherder main view:

**Push Health info**

![treeherder03](https://user-images.githubusercontent.com/3901809/71326609-f419f100-24db-11ea-886e-2cc8e2889a3b.PNG)

--------

**Job info, in red**

![treeherder01](https://user-images.githubusercontent.com/3901809/71326610-f419f100-24db-11ea-9bfe-fd91042b3f8e.PNG)

------

**Green links in push jobs OS**

![treeherder02](https://user-images.githubusercontent.com/3901809/71326611-f4b28780-24db-11ea-8d8a-7735cc407c9f.PNG)

For this one, I would suggest maybe inserting a underline or changing the color, as there are a high density of links in a small space.

sarah-clements commented on ui/css/intermittent-failures.css:
```diff
@@ -70,3 +70,28 @@ ul {
   text-align: left;
 }
+
+/* Graphic overriding style */
```
What does graphic mean? Could you be more specific about what it's overriding?

yogmel replied:
Half a month later I don't remember what it does. So yeah, I should have been more specific haha

sarah-clements commented on ui/css/intermittent-failures.css:
```diff
@@ -70,3 +70,28 @@ ul {
+  opacity: 0.9;
+}
+
+/* ReactTable stlye */
```
typo :)

sarah-clements commented on ui/css/perf.css:
```diff
@@ -37,6 +37,32 @@ h1 {
 }
 
+.btn-info {
+  background-color: #0d7d8f !important;
```
I think it'd be confusing to anyone working on this to use the bootstrap/reactstrap "info" class but then in this style sheet overriding that color. If we're not going to use the default info color prop, can we pass in a class name to the button component instead of using a color prop? 

Also a suggestion for whenever you're overriding framework styles is to have a comment to explain why, so future contributors/developers will be able to know why the decision was made.

yogmel replied:
Sure. In order to do that, I would have to change all the `<Button>` which have the `color="info"` prop (that would be around 14 in all apps). And I think it should also have a guideline in the documentation.

Is that ok?

sarah-clements replied:
yes, that's fine.

sarah-clements commented on ui/css/perf.css:
```diff
@@ -267,6 +293,19 @@ th {
 }
 
+.compare-table .text-success {
+  color: #1e8836 !important;
```
I don't think we'd need to change these green and red colors since we're already using icons with them.

yogmel replied:
I may be looking at the wrong part, but this changes the "untriaged"/"invalid" text in the Compare table, which does not have an icon yet.
I have inserted to show where it is changed:
![image](https://user-images.githubusercontent.com/3901809/72012189-45471600-323a-11ea-9c29-23d324f19b8e.png)

I can insert an icon and keep the color as it was before.

sarah-clements replied:
It also changes the thumbs up icon that appears for certain results in the compare view table [ here](https://treeherder.mozilla.org/perf.html#/compare?originalProject=mozilla-central&originalRevision=adeaa48edc50404d3c77a2faaf496ac89711b8f6&newProject=mozilla-central&newRevision=a4592adb2fe04393c1a5df9487ff88a70eeb0881&framework=1). Click on the "Show only important button".

Visually the shade looks slightly different than the default `bg-success` color used by order of magnitude bar (same color as `text-success`). 

I prefer to create a custom class name to assign to elements where needed instead of making a change like this to all ancestors of an element and overriding a bootstrap style.

sarah-clements replied:
And yes, the icon is oK. I think where it makes sense we should use icons for use with red/green instead of trying to get the shades perfect. For someone with red/green color blindess, I don't think the shade change will make a difference.

sarah-clements commented on ui/css/treeherder-global.css:
```diff
@@ -306,7 +306,7 @@ input[type='checkbox'] {
 
 .bg-lightgray {
-  background-color: lightgray;
+  background-color: whitesmoke;
```
This lightgray is fine according to the color contrast checker: https://webaim.org/resources/contrastchecker/ so lets keep it (plus it looks better by providing contrast against the white background).

sarah-clements commented:
I've commented on IFV and Perfherder color changes; I'll let @camd comment on the Treeherder color changes. @yogmel you'll need to rebase and resolve conflicts before this can be merge.

yogmel commented:
I have sent a new push and have taken some screenshots so it might be easier for code reviewing.

**Some notes:**
- I created `.btn-darker-secondary`, `.btn-darker-info`, `.btn-outline-darker-info` and `.text-darker-info` for color alternatives. When calling the `color` prop, the `btn-[color]` can be passed, instead of `className`;

- There are still a couple of elements that _don't_ pass. I will add a comment on the view that has the not passing element.

------------

![graphs](https://user-images.githubusercontent.com/3901809/72181181-8882bf80-33c7-11ea-885a-11c5f153eea1.jpg)

-------

![compare](https://user-images.githubusercontent.com/3901809/72181196-920c2780-33c7-11ea-9831-dd60882399c5.jpg)

----------

![alerts](https://user-images.githubusercontent.com/3901809/72181197-92a4be00-33c7-11ea-959a-b03c0be88300.jpg)

- The "untriaged" text doesn't pass the contrast score. It has `.text-success` class.

-------------

![health](https://user-images.githubusercontent.com/3901809/72181205-96384500-33c7-11ea-9cea-4a3b4ed3fc07.jpg)

- Hover state doesn't pass, but it is already very light. In order for the hover state to pass, it has to be very similar to one of the row's background color.

sarah-clements commented:
Cool, I think this is a nice way to customize the react strap colors. @ionutgoldan @octavian-negru @airimovici please use this new "darker-info" as a color prop for use with react strap  in lieu of "info" (for better contrast) in lieu of "info".

sarah-clements commented:
>     * The "untriaged" text doesn't pass the contrast score. It has `.text-success` class.
I think this is ok here for this link, since we're also using an icon.

>     * Hover state doesn't pass, but it is already very light. In order for the hover state to pass, it has to be very similar to one of the row's background color.

Do you have a suggestion?

sarah-clements commented:
What else is left to review before this gets merged? Has @camd taken a look at the Treeherder changes?

camd commented:
@yogmel : I've been taking a look at this today and with these color changes, it's pretty hard to tell the difference between red and orange failed jobs.  It's also hard to distinguish green ``success`` jobs from running ones, because the color is so muted down.

Can we iterate on those colors a bit more?

sarah-clements commented:
> @yogmel : I've been taking a look at this today and with these color changes, it's pretty hard to tell the difference between red and orange failed jobs. It's also hard to distinguish green `success` jobs from running ones, because the color is so muted down.
> 
> Can we iterate on those colors a bit more?

I'm hesitant to start changing the green and red colors (`success` and `danger`) because some views will have inconsistent color usage. @yogmel Is the issue that the text is hard to read with those colors due to low contrast, or is it that you can't easily distinguish the red from the green? If it's the latter, I think we should go adapt using red and green colors with icons across the entire site to keep it consistent. Thoughts?

sarah-clements commented on ui/css/perf.css:
```diff
@@ -530,10 +530,10 @@ li.pagination-active.active > button {
 /* ReactTable stlye */
 .ReactTable .-pagination .-btn[disabled] {
-  background-color: rgba(0,0,0,0.03);
+  background-color: rgba(0, 0, 0, 0.03);
```
We should probably move all of these styles (and in intermittent failures CSS file) to the `treeherder-global.css` file in case these libraries are used elsewhere in the future.

sarah-clements commented on ui/css/perf.css:
```diff
@@ -37,6 +37,57 @@ h1 {
   }
 }
 
+.btn-darker-secondary {
```
Move all of these to the `treeherder-global` css file.

ionutgoldan commented:
>@ionutgoldan @octavian-negru @airimovici please use this new "darker-info" as a color prop for use with react strap in lieu of "info" (for better contrast) in lieu of "info".

@sarah-clements you mean in future PRs we'll be working on, right?

sarah-clements commented:
> > @ionutgoldan @octavian-negru @airimovici please use this new "darker-info" as a color prop for use with react strap in lieu of "info" (for better contrast) in lieu of "info".
> 
> @sarah-clements you mean in future PRs we'll be working on, right?

Yes, once this lands. @yogmel will be updating our docs at some point with all of the changes she's made that developers need to know about.

yogmel commented:
> Do you have a suggestion? Go with what is easiest. For a hover state, a simple change should be fine.

I fixed it!

> Is the issue that the text is hard to read with those colors due to low contrast, or is it that you can't easily distinguish the red from the green? If it's the latter, I think we should go adapt using red and green colors with icons across the entire site to keep it consistent. Thoughts?

@sarah-clements All of the changes are due to low contrast. I believe the only red-green issue were [those](https://github.com/mozilla/treeherder/pull/5781#issuecomment-568298662).

I agree with changing the colors to pass the contrast score is not an ideal approach. Maybe a whole revision of the color palette would be necessary. 
We could go and insert icons, as the [Minimum ration](https://developer.mozilla.org/pt-BR/docs/Web/Accessibility/Understanding_WCAG/Perceivable/Color_contrast?utm_source=devtools&utm_medium=a11y-panel-checks-color-contrast) for graphical elements is 3 : 1, while for texts is 4.5 : 1. Currently, the `success` green is 3.71 : 1.

For the time being, I reverted the green and increase the opacity a little bit: `rgba(2, 130, 51, 0.95);` (`treeherder-job-buttons.css:329`).

For the `.revision[data-tags~="backout"] .revision-comment` red (referring to the image below), I would suggest changing it because the text needs to be fully read.

![image](https://user-images.githubusercontent.com/3901809/72637516-449c3700-3940-11ea-930f-6876cef536f6.png)

yogmel commented:
> I've been taking a look at this today and with these color changes, it's pretty hard to tell the difference between red and orange failed jobs. 

@camd I could leave the colors as they are now, only to change the **text color**. Darker hues pass better than white.

![image](https://user-images.githubusercontent.com/3901809/72635152-e882e400-393a-11ea-8eaf-77ac9935f521.png)

Another color combination I think it's conflicting is the Dark and Light Grays. I thought that maybe making it italic would make it more differentiable.
The reason why I can't increase the darkness of the Dark gray color is that it won't become distinguishable with the success green. The Light gray is at the Minimum Ratio contrast.

![image](https://user-images.githubusercontent.com/3901809/72637054-2e41ab80-393f-11ea-9672-00d2406de0bc.png)

yogmel commented:
I will send a push with these changes for feedback along with some icons for Push Health notification, as I talked with @camd on Monday.

yogmel commented:
@sarah-clements @camd I have reviewed the changes and it looks ok. The red we talked about is already passing. If this needs further tweaking, please let me know.

yogmel commented:
@camd I have changed the colors we talked about. It looks like the following:

![Screenshot_2020-01-28 Treeherder User Guide](https://user-images.githubusercontent.com/3901809/73273677-2dbcd680-41c3-11ea-90e0-8f28d6ef9a6a.png)

If you or anyone want any other change or have any feedback, please let me know :)

sarah-clements commented:
A comment about the "Th" text in the buttons. Can you make the the font weight the same for all? I think the slightly bolded that the green success text would look good, and removing the italic for pending jobs.

yogmel commented:
@sarah-clements thanks for reviewing!
About making it all bold, I have a concern - the font-weight can help people with color blindness differentiate it from other colors. I have made all of them bolder, please check the screenshots below:

![Screenshot_2020-01-29  4  autoland](https://user-images.githubusercontent.com/3901809/73361239-52788300-4283-11ea-89f9-d485ab32b842.png)

------------

![Screenshot_2020-01-29  4  autoland(1)](https://user-images.githubusercontent.com/3901809/73361242-54424680-4283-11ea-9adb-6889205ab3fb.png)

You can tell there is a tone difference, but it is very subtle.

sarah-clements commented:
Is the bottom image supposed to be simulation of color blindness? I can't really tell the difference between font weights to be honest. If it's important to distinguish between red and green here then I think we need to figure out something more overt.

camd commented:
I think the idea on the bottom image is that, to a color-blind person, the `success` and `running` jobs would look the same.

What if we made all completed jobs bold, but left `pending` and `running` jobs non-bold?

camd commented:
This PR needs a rebase on `master`, fwiw.

yogmel commented:
> I think the idea on the bottom image is that, to a color-blind person, the `success` and `running` jobs would look the same.

Yes, that is correct. Sorry for not specifying. I have edited that comment and inserted some legends.

> What if we made all completed jobs bold, but left `pending` and `running` jobs non-bold?

This works better IMO.

![image](https://user-images.githubusercontent.com/3901809/73394886-398cc380-42bd-11ea-9c63-c5c13e58b4a5.png)

I didn't find any good examples in the Treeherder app right now, that's why I didn't take any screenshots from it.

sarah-clements commented:
@yogmel Once this is merged I'd like to see a pr updating our docs with some of the changes that developers who work on the app need to know about, such as adding `tag="a"` to dropdown menu items and the new `darker-info` color and other color/style changes you've made (you can just include these important changes for now and add more to the docs later). 

We want to get people to start using these new styles right away once this lands but you might need to do a follow up pr to catch any new code that's been merged that is still using the "info" color prop and other styles, etc that have been changed (no need to open a new bug for it, you can open the pr just with a clear title and brief description).

camd commented:
OK, this is merged.  We can await feedback from folks and iterate as necessary.  Nice work @yogmel !!  Congrats.  This was a huge undertaking.  :)

yogmel commented:
> OK, this is merged. We can await feedback from folks and iterate as necessary. Nice work @yogmel !! Congrats. This was a huge undertaking. :)

Thanks! :)
